gren: 0.4.5 -> 0.6.1

robinheghan opened the pull request:
Context: I'm the lead developer of Gren. This is also my first contribution to nixpkgs, I apologize if I've missed something obvious.

This PR updates the Gren programming language to the latest released version.

Gren has become a little tricky to release since 0.5.0. The Gren compiler now consists of a (Haskell-based) backend and a (Gren/Node-based) frontend.

The frontend is bootstrapped, meaning it's written in Gren.

A release build of the frontend is commited to the official git repository, which makes things rather easy.

For the frontend to use the nix-built backend, a `GREN_BIN` environment variable must be set that points to this binary. If not set, the frontend will download a pre-built backend from github and install in XDG_CACHE. To set GREN_BIN, I've opted to use the wrapProgram utility function.

I've also added `git` to `PATH` using `wrapProgram --suffix`, as Gren's package manager relies on git being installed.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/5512

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/5703

TomaSajt commented:
I added a commit that cleans up the package:

It uses `buildNpmPackage` to not have to generate `node_modules` inside of the nixpkgs tree (which is not allowed).

buildNpmPackage correctly installs the binary as written inside package.json, though the wrapping still needed to be done with makeWrapper in postInstall.

I added extra meta fields.

I moved the `backend` package into `passthru.backend` so that it can be build separately if we want to.

Also, I improved the `update.sh` script.

Could you tell me how the versioning / tagging scheme works for `gren`? There's `blob-{version}` and there's `${version}`.
Which one should we be using?

TomaSajt commented:
If you've reviewed my changes please, squash all of the commits in the PR.

The commit message should be `gren: 0.4.5 -> 0.6.0` (maybe also update the PR title)

robinheghan commented:
> It uses buildNpmPackage to not have to generate node_modules inside of the nixpkgs tree (which is not allowed).

The frontend doesn't have any dependencies, and can run without `node_modules`. There should be no reason to use `buildNpmPackage`.

> Could you tell me how the versioning / tagging scheme works for gren? There's blob-{version} and there's ${version}.
Which one should we be using?

`{version}` is the frontend
`blob-{version}` is the backend

TomaSajt commented:
Alright I'll push another commit

TomaSajt commented:
Oh, sorry, I forgot this PR also adds to you the maintainer list.

In that case, please create two commits instead:
- the first should be `maintainers: add robinheghan` which contains only the change to `maintainer-list.nix`
- the second one should be `gren: 0.4.5 -> 0.6.0` with all other changes.

After that I think we're good to go.

robinheghan committed:
maintainers: add robinheghan

robinheghan committed:
gren: 0.4.5 -> 0.6.1

TomaSajt commented:
Good catch with the shebang of the script!

robinheghan commented:
Done

robinheghan commented:
Are we just waiting for all the checks to complete, or are there other things that needs to be done before this can be merged?

robinheghan commented:
@TomaSajt Updated to Gren 0.6.1 which was released today.

I couldn't get the `nix-update` command of `update.hs` to work, so I removed it, and updated the version and github hash in `package.nix` manually.

Is there anything else that needs to be done before this can be merged?

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/is-it-normal-for-ofborg-github-actions-to-stay-pending-forever/67505/1

robinheghan commented on pkgs/by-name/gr/gren/update.sh:
```diff
@@ -1,22 +1,24 @@
 #!/usr/bin/env nix-shell
-#!nix-shell -i bash -p cabal2nix curl jq
+#!nix-shell -i bash -p cabal2nix curl jq nixfmt-rfc-style
```
I apologize. I wasn't aware that this was supposed to run within the top level nix develop shell.

robinheghan replied:
Do you want me to create a new PR removing the line?
